Plane: Add support for tailsitter aircraft in quadplanes

tridge opened the pull request:
This adds support for a tailsitter frame type in quadplanes. The controls in tailsitter are taken directly from the fixed wing aircraft.
Perhaps the most significant change is the addition of a AP_AHRS_View class, which allows some parts of the code to see a rotated view of the aircraft attitude. This allows the AC_AttitudeControl code to see a view rotated by 90 degrees in pitch from the view seen by the fixed wing code.
This is a work-in-progress PR. Leonard is looking at ways of improving the throttle control in the AP_MotorsTailsitter class.
A video showing this flying in SITL with RealFlight will be uploaded shortly

OXINARF commented on ArduPlane/quadplane.h:
```diff
@@ -74,6 +74,15 @@ class QuadPlane
     // see if we are flying from vtol point of view
     bool is_flying_vtol(void);
+
+    // return true tailsitter frame configured
```
Nitpicking: looks to be missing a when

OXINARF commented on ArduPlane/quadplane.h:
```diff
@@ -74,6 +74,15 @@ class QuadPlane
+    // return true tailsitter frame configured
+    bool is_tailsitter(void);
+
+    // return true when flying a tailsitter in VOTL
```
Nitpicking: should it be VTOL?

tridge commented:
> Wouldn't it be easier if AP_AHRS_View was a subclass of AP_AHRS_NavEKF with a method to change the rotation? I'm pretty sure you thought of that, so can you explain what prevents it?

That would mean duplicating all of the state information inside AP_AHRS_NavEKF, which is quite considerable. I also wanted to do it in a way that didn't add a lot more complexity to AP_AHRS, as it is already pretty complex.

OXINARF commented:
> That would mean duplicating all of the state information inside AP_AHRS_NavEKF, which is quite considerable.

My idea was to then create that subclass in the vehicles instead of AP_AHRS_NavEKF, having the two would be a no-no obviously.

OXINARF commented:
Ah forgot one thing: the file `plane-tailsitter.parm` is missing.

tridge commented:
as a bonus, this PR now also supports dual-motor tailsitters, and as a side effect, supports differential thrust for yaw in fixed wing

magicrub commented:
Dual motor for tail sitters, or dual motors for all? Full yaw support?
Tridge, you're amazing

On Feb 12, 2017 7:50 PM, "Andrew Tridgell" <notifications@github.com> wrote:

> as a bonus, this PR now also supports dual-motor tailsitters, and as a
> side effect, supports differential thrust for yaw in fixed wing
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/5708#issuecomment-279288664>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEj7G-J7xJ3QV80upHCQG6Nal_psjZy7ks5rb9L_gaJpZM4L-WvZ>
> .
>

tridge commented:
yes, full dual-motor yaw support for all fixed wing.
It's only a couple of lines of code now we have the SRV_Channel split done

tridge commented:
I've fixed up all the comments, and removed the todo for heli VFF, as it turns out there is nothing needed for that.
So I think this is ready for master

magicrub commented:
Closes https://github.com/ArduPilot/ardupilot/issues/1861

OXINARF commented:
@tridge Have you missed my comments? 

- the commit named "AC_AttitudeControl: added support for AP_AHRS_View" still has the AHRS_View new files there, they should be in the commit named "AP_AHRS: added support for AP_AHRS_View" instead
- the file plane-tailsitter.parm is still missing

tridge commented:
after successful testing yesterday and today this has been pushed to master

TopGun17 commented:
Helloo!

Nice project tridge! I'm also doing some scool project and i have to built tailsitter/3d hoover plane. Single engine RC plane, same as you have in youtube video.
My question, can i see somewhere the full tailsitter code you showed in video? This would be so helpful to me!

Regards
TopGun

magicrub commented:
>can i see somewhere the full tailsitter code you showed in video

@TopGun17 You're asking for the source on a thread where he's presenting the source :)
https://github.com/ArduPilot/ardupilot/pull/5708/files
